test/e2e/periodic_test.go

```diff
@@ -45,5 +52,76 @@ var _ = Describe("ControlPlaneMachineSet Operator", framework.Periodic(), func()
 			})
 		})
 
+		Context("and an instance is terminated on the cloud provider", func() {
+			BeforeEach(func() {
+				client := testFramework.GetClient()
+				machineList := &machinev1beta1.MachineList{}
+				machineSelector := runtimeclient.MatchingLabels(framework.ControlPlaneMachineSetSelectorLabels())
+
+				By("Getting a list of all control plane machines")
+				Expect(client.List(testFramework.GetContext(), machineList, machineSelector)).To(Succeed(), "should be able to retrieve list of control plane machines")
+
+				By("Deleting an instance from the cloud provider")
+				Expect(testFramework.DeleteAnInstanceFromCloudProvider(&machineList.Items[0])).To(Succeed())
+
+				By("Waiting for the machine to get into Failed phase")
+				Eventually(komega.Object(&machineList.Items[0]), 10*time.Minute).Should(HaveField("Status.Phase", HaveValue(Equal("Failed"))))
+
+				By("Deleting a control plane machine in phase Failed at index 0")
+				Expect(client.Delete(testFramework.GetContext(), &machineList.Items[0])).To(Succeed())
+			})
+
+			helpers.ItShouldReplaceTheOutDatedMachineInDeleting(testFramework, 0)
+		})
+
+		Context("and a node with terminated kubelet", func() {
+			var delObjects map[string]runtimeclient.Object
+			var client runtimeclient.Client
+			var ctx context.Context
+
+			BeforeEach(func() {
+				client = testFramework.GetClient()
+				ctx = testFramework.GetContext()
+				delObjects = make(map[string]runtimeclient.Object)
+				machineList := &machinev1beta1.MachineList{}
+				machineSelector := runtimeclient.MatchingLabels(framework.ControlPlaneMachineSetSelectorLabels())
+				node := &corev1.Node{}
+
+				By("Getting a list of all control plane machines")
+				Expect(client.List(ctx, machineList, machineSelector)).To(Succeed(), "should be able to retrieve list of control plane machines")
+
+				By("Getting the node from the machine")
+				Expect(client.Get(ctx, types.NamespacedName{
+					Namespace: framework.MachineAPINamespace,
+					Name:      machineList.Items[0].Status.NodeRef.Name,
+				}, node)).To(Succeed())
+
+				By("Shutting down the kubelet on a node")
+				Expect(testFramework.TerminateKubelet(node, delObjects)).To(Succeed())
+
+				By("Waiting for the node to get into NotReady phase")
+				Eventually(komega.Object(node), 10*time.Minute).Should(
+					WithTransform(func(node *corev1.Node) []corev1.ConditionStatus {
+						statuses := []corev1.ConditionStatus{}
+						for _, condition := range node.Status.Conditions {
+							statuses = append(statuses, condition.Status)
+						}
+
+						return statuses
+					}, ContainElements(corev1.ConditionUnknown, corev1.ConditionFalse)))
```

Comment:
What exactly are you checking here? To fit with the rest of the tests here, we would just use gomega constructs and be explicit about which conditions should have which statuses
```suggestion
				Eventually(komega.Object(node), 10*time.Minute).Should(HaveField("Status.Conditions", 
					ContainElements(
						SatisfyAll(HaveField("Type", Equal(<something>),HaveField("Status", Equal(corev1.ConditionUnknown))),
						SatisfyAll(HaveField("Type", Equal(<something>),HaveField("Status", Equal(corev1.ConditionFalse))),
					),
				))
```

Reply:
Is this actually correct?

We have `nodeObj.Status.Conditions` as a list of `NodeCondition` and we need to check that every `NodeCondition.Status` is either `Unknown` or `False`. What didn't you like about my solution? Not idiomatic? Thanks for clarifying!

Reply:
I wasn't actually sure what you were trying to achieve with that check, but, if you wanted to check that each condition was Unknown or False, your current approach also won't work, since it would pass for `{Unknown,False,True,True}`

If you want to make sure each one is either Unknown or False, then you need
```suggestion
				Eventually(komega.Object(node), 10*time.Minute).Should(HaveField("Status.Conditions", 
					HaveEach(HaveField("Status",
						SatisfyAny(Equal(corev1.ConditionUnknown), Equal(corev1.ConditionFalse)),
					)),
				))
```
This checks that for every element in the list, the Status field is Unknown or False